### PR TITLE
Add ragged workspace support for binary operations 

### PR DESCRIPTION
**Pull request by rosswhitfield**
### Description of work

You should now be added to do binary operations with Ragged Workspaces

#### Summary of work

Fixes [4790: Math operations on mantid ragged workspaces](https://ornlrse.clm.ibmcloud.com/ccm/resource/itemName/com.ibm.team.workitem.WorkItem/4790)

### To test:

The are 4 algorithms to test, Plus, Minus, Divide and Multiply. And you should test with EventWorkspace and Histogram workspaces, also with a constant (_e.g._ `ws+2`).

#### real data example

```python
filename="SNAP_45874"  # from systemtest data
LoadEventNexus(Filename=filename, OutputWorkspace='left')
LoadEventNexus(Filename=filename, OutputWorkspace='right')

CreateGroupingWorkspace(InputWorkspace='left', GroupDetectorsBy='Column', OutputWorkspace='SNAPFocGrp_Column')
for name in 'left', 'right':
    ConvertUnits(InputWorkspace=name, OutputWorkspace=name, Target='dSpacing')
    DiffractionFocussing(InputWorkspace=name, OutputWorkspace=name, GroupingWorkspace='SNAPFocGrp_Column')
    RebinRagged(InputWorkspace=name, OutputWorkspace=name+"_hist", XMin=(0.5,0.6,0.7,0.8,0.9,1), XMax=(2,2.2,2.4,2.6,2.8,3), Delta=(-0.001,-0.0015,-0.002,-0.0025,-0.003,-0.0035), PreserveEvents=False)
    RebinRagged(InputWorkspace=name, OutputWorkspace=name, XMin=(0.5,0.6,0.7,0.8,0.9,1), XMax=(2,2.2,2.4,2.6,2.8,3), Delta=(-0.001,-0.0015,-0.002,-0.0025,-0.003,-0.0035), PreserveEvents=True)

divisor = mtd['left']/mtd['right']
minusor = mtd['left'] - mtd['right']
```

#### fake data example

```python
left = CreateSampleWorkspace("Event", NumBanks=1, BankPixelWidth=2)
right = CreateSampleWorkspace("Event", NumBanks=1, BankPixelWidth=2)

left=RebinRagged(InputWorkspace=left, XMin=(100,200,300,400), XMax=(20000,15000,10000,5000), Delta=-1, PreserveEvents=True)
right=RebinRagged(InputWorkspace=right, XMin=(100,200,300,400), XMax=(20000,15000,10000,5000), Delta=-1, PreserveEvents=True)

out = left + right
out = left * right
out = left * 5
left =/ 10
# and many more combinations to try
```

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

File: Framework/Algorithms/src/BinaryOperation.cpp [hunk 1]
```diff
@@ -372,8 +380,8 @@ std::string BinaryOperation::checkSizeCompatibility(const API::MatrixWorkspace_c
   if (m_rhsBlocksize == 1 && lhs->getNumberHistograms() == rhs->getNumberHistograms())
     return "";
   // Past this point, we require the X arrays to match. Note this only checks
-  // the first spectrum
-  if (!WorkspaceHelpers::matchingBins(*lhs, *rhs, true)) {
+  // the first spectrum except for ragged workspaces
+  if (!WorkspaceHelpers::matchingBins(*lhs, *rhs, !m_lhsRagged && !m_rhsRagged)) {
```
Comment (peterfpeterson) [inline Framework/Algorithms/src/BinaryOperation.cpp:384]: I like the way this is done. Very clear

File: Framework/Algorithms/test/PlusMinusTest.in.h [hunk 2]
```diff
@@ -783,12 +788,21 @@ class @PLUSMINUSTEST_CLASS@ : public CxxTest::TestSuite
   void checkData( MatrixWorkspace_sptr work_in1,  MatrixWorkspace_sptr work_in2, MatrixWorkspace_sptr work_out1,
       int loopOrientation, double expectedValue=-1.0, double expectedError=-1.0  )
   {
+    std::cout << "hello world!\n";
```
Comment (peterfpeterson) [inline Framework/Algorithms/test/PlusMinusTest.in.h:791]: Remove this debug print
```suggestion
```

**Review by peterfpeterson — COMMENTED** [review]
Only minor comments from looking through the code. I haven't tried it out yet.

File: Framework/Algorithms/src/BinaryOperation.cpp [hunk 3]
```diff
@@ -134,8 +134,16 @@ void BinaryOperation::exec() {
   m_rhs = getProperty(inputPropName2());
   m_AllowDifferentNumberSpectra = getProperty("AllowDifferentNumberSpectra");
 
-  m_lhsBlocksize = m_lhs->blocksize();
-  m_rhsBlocksize = m_rhs->blocksize();
+  try {
```
[the remark this answers is not in the document]
Comment (rosswhitfield) [reply Framework/Algorithms/src/BinaryOperation.cpp:137]: If you look at the implementation of `blocksize` and `isRaggedWorkspace`, they basically run the same code just looking for the opposite result. So I though doing `isRaggedWorkspace` after `blocksize` was redundant.

**Review by peterfpeterson — APPROVED** [review]
It appears to work as expected

**Pull request merged**